rockets/jsonrpc/receiver.cpp

```diff
@@ -153,45 +157,91 @@ class Receiver::Impl
     }
 
     void processCommand(const json& request, const uintptr_t clientID,
-                        std::function<void(json)> callback)
+                        JsonResponseCallback respond)
     {
         const auto id = request.count("id") ? request["id"] : json();
         const bool isNotification = id.is_null();
         if (!_isValidJsonRpcRequest(request))
         {
             if (isNotification)
-                callback(json());
+                respond(json());
             else
-                callback(makeErrorResponse(invalidRequest, id));
+                respond(makeErrorResponse(invalidRequest, id));
             return;
         }
 
         const auto methodName = request["method"].get<std::string>();
+
+        if(methodName == cancelMethodName)
+        {
+            processCancel(request, respond);
+            return;
+        }
+
         const auto method = methods.find(methodName);
         if (method == methods.end())
         {
             if (isNotification)
-                callback(json());
+                respond(json());
             else
-                callback(makeErrorResponse(methodNotFound, id));
+                respond(makeErrorResponse(methodNotFound, id));
             return;
         }
 
+        const auto& cancelFunc = cancels.find(methodName);
+        if(cancelFunc != cancels.end())
+            pendingRequests[id] = {cancelFunc->second, respond};
+
         const auto params = request.find("params") == request.end()
                                 ? ""
                                 : dump(request["params"]);
         const auto& func = method->second;
         func({params, clientID},
-             [callback, id](const Response rep) {
+             [respond, id, &pendingRequests = pendingRequests](const Response rep) {
```

Comment:
there's a risk capturing pendingRequests by reference, it might lead to a bad access if *this* is destroyed before the response arrives

Reply:
Hmm, that's bad. So no reliable place to clear this map after the request has been handled?

Reply:
I'm not sure, we can take a look tomorrow. One solution could be to use a shared_ptr to store the map and give a weak_ptr to the callback.